sdk/core/Azure.Core.Experimental/tests/public/JsonDataPublicTests.cs

```diff
@@ -0,0 +1,400 @@
+        public void DynamicCanConvertToIEnumerableInt()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            int i = 0;
+            foreach (int dynamicItem in jsonData)
+            {
+                Assert.AreEqual(i, dynamicItem);
+
+                i++;
+            }
+            Assert.AreEqual(4, i);
+        }
+
+        [Test]
+        public void DynamicArrayHasLength()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            Assert.AreEqual(4, jsonData.Length);
+        }
+
+        [Test]
+        public void DynamicArrayFor()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            for (int i = 0; i < jsonData.Length; i++)
+            {
+                Assert.AreEqual(i, (int)jsonData[i]);
+            }
+        }
+
+        [Test]
+        public void CanAccessProperties()
+        {
+            dynamic jsonData = new BinaryData("{ \"primitive\":\"Hello\", \"nested\": { \"nestedPrimitive\":true } }").ToDynamic();
+
+            Assert.AreEqual("Hello", (string)jsonData.primitive);
+            Assert.AreEqual(true, (bool)jsonData.nested.nestedPrimitive);
+        }
+
+        [Test]
+        public void CanTestPropertyForNull()
+        {
+            dynamic jsonData = new BinaryData("{ \"primitive\":\"Hello\", \"nested\": { \"nestedPrimitive\":true } }").ToDynamic();
+
+            Assert.IsNull(jsonData.OptionalInt);
+            Assert.IsNull(jsonData.OptionalString);
+            Assert.AreEqual("Hello", (string)jsonData.primitive);
+        }
+
+        [Test]
+        public void CanAddStringToList()
+        {
+            dynamic jsonData = new BinaryData(new { value = "foo" }).ToDynamic();
+
+            List<string> list = new();
+
+            // TODO: Should we add an implicit cast to string?
+            // Can we do it dynamically?
+            // Alternatively, we could provide a different type for string leaf
+            // nodes, like we do for numbers.
+            list.Add(jsonData.value);
+
+            Assert.AreEqual(1, list.Count);
+            Assert.AreEqual("foo", list[0]);
+        }
+
+        [Test]
+        public void CanAddIntToList()
+        {
+            dynamic jsonData = new BinaryData(new { value = 5 }).ToDynamic();
+
+            List<int> list = new();
+
+            // TODO: Should we add an implicit cast to int?
+            // Can we do it dynamically?
+            list.Add(jsonData.value);
+
+            Assert.AreEqual(1, list.Count);
+            Assert.AreEqual(5, list[0]);
+        }
+
+        [Test]
+        public void GetMemberIsCaseInsensitive()
+        {
+            dynamic jsonData = new BinaryData("{ \"primitive\":\"Hello\", \"nested\": { \"nestedPrimitive\":true } }").ToDynamic();
+
+            Assert.AreEqual("Hello", (string)jsonData.Primitive);
+            Assert.AreEqual(true, (bool)jsonData.Nested.NestedPrimitive);
```

Comment:
If we want to make this configurable (see discussion [here](https://github.com/Azure/azure-sdk-for-net/pull/31769#discussion_r1017018043)), we would probably expose this as letting callers pass an options type to BinaryData's ToDynamic() method.  This raises several questions, for example, do we want to create an options type for JsonData/DynamicData to control this, or would we want to piggyback on JsonSerializerOptions PropertyNameCaseInsensitive property?